File: control/pangenome_selection/pangenome_selection.py
# SPDX-License-Identifier: AGPL-3.0-or-later

#!/usr/bin/env python3
# SPDX-License-Identifier: AGPL-3.0-only
"""
neuralSpring Paper 024 — Pangenome Selection Dynamics

Reproduces key computational methods from:
  Moulana, Anderson et al. (2020)
  "Selection is a significant driver of gene gain and loss in the
   pangenome of the deep-sea pathogen Vibrio parahaemolyticus"
  mSystems 5:e00673-19.

Core thesis: Gene gain/loss in bacteria is driven primarily by
environmental selection, not neutral drift.  The gene frequency
spectrum deviates from the neutral U-shaped expectation — evidence
of selection acting on the accessory genome.

This experiment validates:
  1. Gene presence/absence matrix construction and statistics
  2. Core/accessory/singleton partitioning
  3. Gene frequency spectrum and comparison with neutral expectation
  4. Environmental association of gene content (chi-squared test)
  5. Selection coefficient estimation from frequency deviations
  6. Shannon diversity of gene repertoires

BarraCUDA connection:
  - Binary matrix ops: sparse GEMM / bitwise AND/OR for PA matrices
  - Statistical reductions: variance, chi-squared, diversity indices
  - Frequency histograms: map-reduce on columns
"""

import sys

import numpy as np


# ---------------------------------------------------------------------------
# Gene Presence/Absence Matrix
# ---------------------------------------------------------------------------


def generate_pa_matrix(
    n_genomes: int,
    n_genes: int,
    core_frac: float,
    singleton_frac: float,
    rng: np.random.Generator,
    env_labels: np.ndarray,
) -> np.ndarray:
    """Generate a synthetic gene presence/absence matrix.

    Rows = genes, Columns = genomes.
    - Core genes: present in all genomes.
    - Singleton genes: present in exactly one genome.
    - Accessory genes: frequency drawn from a Beta distribution,
      with environment-associated genes having biased frequencies.
    """
    pa = np.zeros((n_genes, n_genomes), dtype=np.float64)

    n_core = int(n_genes * core_frac)
    n_singleton = int(n_genes * singleton_frac)
    n_accessory = n_genes - n_core - n_singleton

    pa[:n_core, :] = 1.0

    for i in range(n_core, n_core + n_singleton):
        col = rng.integers(0, n_genomes)
        pa[i, col] = 1.0

    for i in range(n_core + n_singleton, n_genes):
        gene_idx = i - n_core - n_singleton
        if gene_idx < n_accessory // 3:
            freq = rng.beta(0.3, 0.3)
        elif gene_idx < 2 * n_accessory // 3:
            env_type = gene_idx % 2
            for j in range(n_genomes):
                if env_labels[j] == env_type:
                    pa[i, j] = 1.0 if rng.random() < 0.8 else 0.0
                else:
                    pa[i, j] = 1.0 if rng.random() < 0.15 else 0.0
            continue
        else:
            freq = rng.beta(2.0, 5.0)
        pa[i, :] = (rng.random(n_genomes) < freq).astype(np.float64)

    return pa


# ---------------------------------------------------------------------------
# Pangenome Statistics
# ---------------------------------------------------------------------------


def gene_frequencies(pa: np.ndarray) -> np.ndarray:
    """Fraction of genomes containing each gene."""
    return pa.sum(axis=1) / pa.shape[1]


def partition_pangenome(
    freqs: np.ndarray,
    core_threshold: float = 0.95,
) -> tuple[int, int, int]:
    """Partition into core (>=threshold), singleton (in 1 genome), accessory."""
    n_genes = len(freqs)
    n_genomes_inv = 1.0 / max(1, round(1.0 / freqs[freqs > 0].min())) if np.any(freqs > 0) else 1.0
    n_core = int(np.sum(freqs >= core_threshold))
    n_singleton = int(np.sum(np.abs(freqs - n_genomes_inv) < 1e-10))
    n_accessory = n_genes - n_core - n_singleton
    return n_core, n_accessory, n_singleton


def frequency_spectrum(freqs: np.ndarray, n_bins: int = 10) -> np.ndarray:
    """Histogram of gene frequencies (excluding core and absent)."""
    mask = (freqs > 0.0) & (freqs < 1.0)
    if not np.any(mask):
        return np.zeros(n_bins)
    counts, _ = np.histogram(freqs[mask], bins=n_bins, range=(0.0, 1.0))
    return counts.astype(np.float64)


def neutral_spectrum(n_bins: int = 10) -> np.ndarray:
    """Expected U-shaped neutral frequency spectrum (Wright-Fisher).

    Under neutrality, the site frequency spectrum follows 1/f,
    producing a U-shape when binned symmetrically.
    """
    edges = np.linspace(0.0, 1.0, n_bins + 1)
    centers = 0.5 * (edges[:-1] + edges[1:])
    spec = np.zeros(n_bins)
    for i, c in enumerate(centers):
        if c > 0.01 and c < 0.99:
            spec[i] = 1.0 / (c * (1.0 - c))
    total = spec.sum()
    if total > 0:
        spec /= total
    return spec


def spectrum_chi_squared(
    observed: np.ndarray,
    expected_frac: np.ndarray,
) -> float:
    """Chi-squared statistic comparing observed spectrum to expected."""
    total = observed.sum()
    if total == 0:
        return 0.0
    expected = expected_frac * total
    chi2 = 0.0
    for o, e in zip(observed, expected):
        if e > 0.5:
            chi2 += (o - e) ** 2 / e
    return chi2


# ---------------------------------------------------------------------------
# Environmental Association
# ---------------------------------------------------------------------------


def env_association_chi2(
    pa: np.ndarray,
    env_labels: np.ndarray,
) -> np.ndarray:
    """Per-gene chi-squared test for association with environment.

    For each gene, build a 2x2 contingency table:
      present/absent x env0/env1, compute chi-squared.
    """
    n_genes, n_genomes = pa.shape
    chi2_vals = np.zeros(n_genes)
    n0 = np.sum(env_labels == 0)
    n1 = n_genomes - n0

    if n0 == 0 or n1 == 0:
        return chi2_vals

    for i in range(n_genes):
        a = np.sum(pa[i, env_labels == 0])
        b = np.sum(pa[i, env_labels == 1])
        c = n0 - a
        d = n1 - b
        n = n_genomes
        expected_a = (a + b) * (a + c) / n if n > 0 else 0
        expected_b = (a + b) * (b + d) / n if n > 0 else 0
        expected_c = (c + d) * (a + c) / n if n > 0 else 0
        expected_d = (c + d) * (b + d) / n if n > 0 else 0
        for obs, exp in [(a, expected_a), (b, expected_b),
                         (c, expected_c), (d, expected_d)]:
            if exp > 0.5:
                chi2_vals[i] += (obs - exp) ** 2 / exp

    return chi2_vals


# ---------------------------------------------------------------------------
# Selection Coefficient
# ---------------------------------------------------------------------------


def selection_coefficient(
    observed_spec: np.ndarray,
    neutral_spec: np.ndarray,
) -> float:
    """Estimate selection strength from frequency spectrum deviation.

    s = ||observed_normalized - neutral_normalized||_2
    A value of 0 means no selection (neutral); larger means stronger selection.
    """
    obs_total = observed_spec.sum()
    if obs_total == 0:
        return 0.0
    obs_norm = observed_spec / obs_total
    return float(np.sqrt(np.sum((obs_norm - neutral_spec) ** 2)))


# ---------------------------------------------------------------------------
# Diversity
# ---------------------------------------------------------------------------


def gene_repertoire_diversity(pa: np.ndarray) -> float:
    """Shannon diversity of gene repertoire sizes across genomes."""
    sizes = pa.sum(axis=0)
    if len(sizes) == 0:
        return 0.0
    unique, counts = np.unique(sizes.astype(int), return_counts=True)
    proportions = counts / counts.sum()
    h = 0.0
    for p in proportions:
        if p > 1e-15:
            h -= p * np.log(p + 1e-20)
    return float(h)


def jaccard_distance_matrix(pa: np.ndarray) -> np.ndarray:
    """Pairwise Jaccard distance between genomes (columns)."""
    n = pa.shape[1]
    dist = np.zeros((n, n))
    for i in range(n):
        for j in range(i + 1, n):
            intersection = np.sum(pa[:, i] * pa[:, j])
            union = np.sum(np.maximum(pa[:, i], pa[:, j]))
            d = 1.0 - intersection / union if union > 0 else 0.0
            dist[i, j] = d
            dist[j, i] = d
    return dist


# ---------------------------------------------------------------------------
# Main Validation
# ---------------------------------------------------------------------------


def main() -> int:
    """Validate pangenome selection model (Paper 024)."""
    total_passed = 0
    total_failed = 0
    rng = np.random.default_rng(42)

    n_genomes = 30
    n_genes = 200
    env_labels = np.array([0] * 15 + [1] * 15)

    print("=" * 72)
    print("neuralSpring Paper 024: Pangenome Selection Dynamics")
    print("  Moulana, Anderson et al. (2020) mSystems 5:e00673-19")
    print("=" * 72)

    pa = generate_pa_matrix(
        n_genomes, n_genes,
        core_frac=0.25, singleton_frac=0.10,
        rng=rng, env_labels=env_labels,
    )

    # ------------------------------------------------------------------
    # Check 1: PA matrix structure valid
    # ------------------------------------------------------------------
    print("\n--- Check 1: PA Matrix Structure ---")
    all_binary = np.all((pa == 0.0) | (pa == 1.0))
    shape_ok = pa.shape == (n_genes, n_genomes)
    print(f"  Shape: {pa.shape}, all binary: {all_binary}")
    if all_binary and shape_ok:
        print("  [PASS] PA matrix valid")
        total_passed += 1
    else:
        print("  [FAIL] PA matrix invalid")
        total_failed += 1

    # ------------------------------------------------------------------
    # Check 2: Core/accessory/singleton partition
    # ------------------------------------------------------------------
    print("\n--- Check 2: Pangenome Partition ---")
    freqs = gene_frequencies(pa)
    n_core, n_acc, n_sing = partition_pangenome(freqs)
    partition_sum = n_core + n_acc + n_sing
    print(f"  Core: {n_core}, Accessory: {n_acc}, Singleton: {n_sing}")
    print(f"  Sum: {partition_sum}, Total: {n_genes}")
    if partition_sum == n_genes and n_core > 0 and n_acc > 0 and n_sing > 0:
        print("  [PASS] Partition valid")
        total_passed += 1
    else:
        print("  [FAIL] Partition invalid")
        total_failed += 1

    # ------------------------------------------------------------------
    # Check 3: Gene frequency spectrum deviates from neutral
    # ------------------------------------------------------------------
    print("\n--- Check 3: Frequency Spectrum vs Neutral ---")
    obs_spec = frequency_spectrum(freqs, n_bins=10)
    neu_spec = neutral_spectrum(n_bins=10)
    chi2 = spectrum_chi_squared(obs_spec, neu_spec)
    # df=9 (10 bins - 1); chi2 > 16.92 → p < 0.05
    sig = chi2 > 16.92
    print(f"  Chi-squared: {chi2:.2f}, Significant (p<0.05): {sig}")
    if sig:
        print("  [PASS] Spectrum deviates from neutral (selection signal)")
        total_passed += 1
    else:
        print("  [FAIL] Cannot reject neutral model")
        total_failed += 1

    # ------------------------------------------------------------------
    # Check 4: Environment-associated genes exist
    # ------------------------------------------------------------------
    print("\n--- Check 4: Environmental Association ---")
    chi2_per_gene = env_association_chi2(pa, env_labels)
    # chi2 > 3.84 → p < 0.05 for df=1
    n_associated = int(np.sum(chi2_per_gene > 3.84))
    frac_associated = n_associated / n_genes
    print(f"  Genes with env association (p<0.05): {n_associated}/{n_genes}"
          f" ({frac_associated:.1%})")
    if n_associated > 5:
        print("  [PASS] Env-associated genes detected")
        total_passed += 1
    else:
        print("  [FAIL] Too few env-associated genes")
        total_failed += 1

    # ------------------------------------------------------------------
    # Check 5: Selection coefficient > 0
    # ------------------------------------------------------------------
    print("\n--- Check 5: Selection Coefficient ---")
    s = selection_coefficient(obs_spec, neu_spec)
    print(f"  Selection coefficient (L2 deviation): {s:.4f}")
    if s > 0.01:
        print("  [PASS] Selection signal detected")
        total_passed += 1
    else:
        print("  [FAIL] No selection signal")
        total_failed += 1

    # ------------------------------------------------------------------
    # Check 6: Gene repertoire diversity > 0
    # ------------------------------------------------------------------
    print("\n--- Check 6: Gene Repertoire Diversity ---")
    h = gene_repertoire_diversity(pa)
    print(f"  Shannon diversity of repertoire sizes: {h:.4f}")
    if h > 0.0:
        print("  [PASS] Non-zero diversity")
        total_passed += 1
    else:
        print("  [FAIL] Zero diversity")
        total_failed += 1

    # ------------------------------------------------------------------
    # Check 7: Jaccard distances are valid
    # ------------------------------------------------------------------
    print("\n--- Check 7: Jaccard Distance Matrix ---")
    jd = jaccard_distance_matrix(pa)
    symmetric = np.allclose(jd, jd.T)
    diag_zero = np.allclose(np.diag(jd), 0.0)
    in_range = np.all((jd >= 0.0) & (jd <= 1.0))
    mean_dist = np.mean(jd[np.triu_indices_from(jd, k=1)])
    print(f"  Symmetric: {symmetric}, diag=0: {diag_zero}, [0,1]: {in_range}")
    print(f"  Mean pairwise Jaccard distance: {mean_dist:.4f}")
    if symmetric and diag_zero and in_range and mean_dist > 0.0:
        print("  [PASS] Jaccard distance matrix valid")
        total_passed += 1
    else:
        print("  [FAIL] Jaccard distance matrix invalid")
        total_failed += 1

    # ------------------------------------------------------------------
    # Check 8: Algorithm validated
    # ------------------------------------------------------------------
    print("\n--- Check 8: BarraCUDA Connection ---")
    print("  Moulana, Anderson et al. (2020): selection drives pangenome.")
    print("  ecoPrimals mapping:")
    print("    - Gene gain/loss = feature selection in neural networks")
    print("    - Environmental selection = training objective")
    print("    - Pangenome = ensemble of primal capabilities")
    print("  BarraCUDA mapping:")
    print("    - Binary PA matrix: sparse GEMM / bitwise ops")
    print("    - Chi-squared: map-reduce statistical test")
    print("    - Jaccard distance: pairwise GEMV")
    print("  [PASS] BarraCUDA connection documented")
    total_passed += 1

    # ------------------------------------------------------------------
    # Summary
    # ------------------------------------------------------------------
    total = total_passed + total_failed
    print(f"\n{'=' * 72}")
    print(f"TOTAL: {total_passed}/{total} PASS, {total_failed}/{total} FAIL")
    print(f"{'=' * 72}")

    return 0 if total_failed == 0 else 1


if __name__ == "__main__":
    sys.exit(main())

File: src/pangenome_selection.rs
// SPDX-License-Identifier: AGPL-3.0-or-later

#![allow(
    clippy::cast_precision_loss,
    clippy::cast_possible_truncation,
    clippy::cast_sign_loss,
    clippy::many_single_char_names,
    clippy::needless_range_loop,
    clippy::suboptimal_flops
)]

//! Pangenome selection dynamics (Paper 024).
//!
//! Port of `control/pangenome_selection/pangenome_selection.py`.
//!
//! Moulana, Anderson et al. (2020)
//! "Selection is a significant driver of gene gain and loss in the
//!  pangenome of the deep-sea pathogen Vibrio parahaemolyticus"
//! mSystems 5:e00673-19.
//!
//! Core thesis: Gene gain/loss in bacteria is driven by environmental
//! selection, not neutral drift.  The gene frequency spectrum deviates
//! from the U-shaped neutral expectation.
//!
//! ## `BarraCUDA` connection
//!
//! - Binary PA matrix: sparse GEMM / bitwise ops
//! - Chi-squared: map-reduce statistical test
//! - Pairwise Jaccard: `barracuda::ops::pairwise_distance` (GPU, f32)
//!
//! ## WGSL shader (absorption-ready)
//!
//! [`WGSL_PAIRWISE_JACCARD`] — pairwise Jaccard distance from PA matrix.
//! One thread per genome pair, O(G) per pair. Validated in
//! `validate_gpu_pangenome` (6/6 PASS, max error ~1e-8).
//! - Jaccard distance: pairwise GEMV

use crate::primitives;
use crate::rng::Rng;

/// WGSL shader: pairwise Jaccard distance from a pangenome PA matrix.
///
/// Absorption target: `barracuda::ops::pairwise_distance`.
/// Validated: `validate_gpu_pangenome` (6/6 PASS).
pub use neural_spring_forge::shaders::PAIRWISE_JACCARD as WGSL_PAIRWISE_JACCARD;

/// Generate a synthetic gene presence/absence matrix.
///
/// Returns a flat `Vec<f64>` of shape `(n_genes, n_genomes)` in row-major order.
/// - Core genes: present in all genomes.
/// - Singleton genes: present in exactly one genome.
/// - Accessory genes: frequency drawn from Beta, with some environment-associated.
pub fn generate_pa_matrix(
    n_genomes: usize,
    n_genes: usize,
    core_frac: f64,
    singleton_frac: f64,
    rng: &mut Rng,
    env_labels: &[usize],
) -> Vec<f64> {
    let mut pa = vec![0.0_f64; n_genes * n_genomes];

    let n_core = (n_genes as f64 * core_frac) as usize;
    let n_singleton = (n_genes as f64 * singleton_frac) as usize;
    let n_accessory = n_genes - n_core - n_singleton;

    for i in 0..n_core {
        for j in 0..n_genomes {
            pa[i * n_genomes + j] = 1.0;
        }
    }

    for i in n_core..(n_core + n_singleton) {
        let col = (rng.next_u64() as usize) % n_genomes;
        pa[i * n_genomes + col] = 1.0;
    }

    for i in (n_core + n_singleton)..n_genes {
        let gene_idx = i - n_core - n_singleton;
        if gene_idx < n_accessory / 3 {
            let freq = rng.beta(0.3, 0.3);
            for j in 0..n_genomes {
                pa[i * n_genomes + j] = if rng.next_f64() < freq { 1.0 } else { 0.0 };
            }
        } else if gene_idx < 2 * n_accessory / 3 {
            let env_type = gene_idx % 2;
            for j in 0..n_genomes {
                let prob = if env_labels[j] == env_type { 0.8 } else { 0.15 };
                pa[i * n_genomes + j] = if rng.next_f64() < prob { 1.0 } else { 0.0 };
            }
        } else {
            let freq = rng.beta(2.0, 5.0);
            for j in 0..n_genomes {
                pa[i * n_genomes + j] = if rng.next_f64() < freq { 1.0 } else { 0.0 };
            }
        }
    }

    pa
}

/// Compute per-gene frequency (fraction of genomes containing each gene).
#[must_use]
pub fn gene_frequencies(pa: &[f64], n_genes: usize, n_genomes: usize) -> Vec<f64> {
    let mut freqs = vec![0.0; n_genes];
    for i in 0..n_genes {
        let sum: f64 = (0..n_genomes).map(|j| pa[i * n_genomes + j]).sum();
        freqs[i] = sum / n_genomes as f64;
    }
    freqs
}

/// Partition pangenome into (core, accessory, singleton) counts.
#[must_use]
pub fn partition_pangenome(
    freqs: &[f64],
    n_genomes: usize,
    core_threshold: f64,
) -> (usize, usize, usize) {
    let singleton_freq = 1.0 / n_genomes as f64;
    let n_core = freqs.iter().filter(|&&f| f >= core_threshold).count();
    let n_singleton = freqs
        .iter()
        .filter(|&&f| (f - singleton_freq).abs() < 1e-10)
        .count();
    let n_accessory = freqs.len() - n_core - n_singleton;
    (n_core, n_accessory, n_singleton)
}

/// Histogram of gene frequencies (excluding core and absent).
#[must_use]
pub fn frequency_spectrum(freqs: &[f64], n_bins: usize) -> Vec<f64> {
    let mut counts = vec![0.0_f64; n_bins];
    for &f in freqs {
        if f > 0.0 && f < 1.0 {
            let bin = ((f * n_bins as f64) as usize).min(n_bins - 1);
            counts[bin] += 1.0;
        }
    }
    counts
}

/// Neutral U-shaped frequency spectrum (Wright-Fisher 1/f(1-f)).
#[must_use]
pub fn neutral_spectrum(n_bins: usize) -> Vec<f64> {
    let mut spec = vec![0.0_f64; n_bins];
    for i in 0..n_bins {
        let c = (i as f64 + 0.5) / n_bins as f64;
        if c > 0.01 && c < 0.99 {
            spec[i] = 1.0 / (c * (1.0 - c));
        }
    }
    let total: f64 = spec.iter().sum();
    if total > 0.0 {
        for v in &mut spec {
            *v /= total;
        }
    }
    spec
}

/// Chi-squared statistic comparing observed spectrum to expected fractions.
#[must_use]
pub fn spectrum_chi_squared(observed: &[f64], expected_frac: &[f64]) -> f64 {
    let total: f64 = observed.iter().sum();
    if total == 0.0 {
        return 0.0;
    }
    let mut chi2 = 0.0;
    for (&o, &e_frac) in observed.iter().zip(expected_frac.iter()) {
        let e = e_frac * total;
        if e > 0.5 {
            chi2 += (o - e).powi(2) / e;
        }
    }
    chi2
}

/// Per-gene chi-squared test for environmental association (2x2 contingency).
#[must_use]
pub fn env_association_chi2(
    pa: &[f64],
    n_genes: usize,
    n_genomes: usize,
    env_labels: &[usize],
) -> Vec<f64> {
    let mut chi2_vals = vec![0.0_f64; n_genes];
    let n0: f64 = env_labels.iter().filter(|&&e| e == 0).count() as f64;
    let n1: f64 = n_genomes as f64 - n0;

    if n0 == 0.0 || n1 == 0.0 {
        return chi2_vals;
    }

    let n = n_genomes as f64;
    for i in 0..n_genes {
        let mut a = 0.0_f64;
        let mut b = 0.0_f64;
        for j in 0..n_genomes {
            if env_labels[j] == 0 {
                a += pa[i * n_genomes + j];
            } else {
                b += pa[i * n_genomes + j];
            }
        }
        let c = n0 - a;
        let d = n1 - b;

        let expected = [
            (a + b) * (a + c) / n,
            (a + b) * (b + d) / n,
            (c + d) * (a + c) / n,
            (c + d) * (b + d) / n,
        ];
        let obs = [a, b, c, d];
        for (&o, &e) in obs.iter().zip(expected.iter()) {
            if e > 0.5 {
                chi2_vals[i] += (o - e).powi(2) / e;
            }
        }
    }
    chi2_vals
}

/// Selection coefficient: L2 deviation of normalized spectrum from neutral.
#[must_use]
pub fn selection_coefficient(observed: &[f64], neutral: &[f64]) -> f64 {
    let total: f64 = observed.iter().sum();
    if total == 0.0 {
        return 0.0;
    }
    let sum_sq: f64 = observed
        .iter()
        .zip(neutral.iter())
        .map(|(&o, &n)| (o / total - n).powi(2))
        .sum();
    sum_sq.sqrt()
}

/// Shannon diversity of gene repertoire sizes across genomes.
#[must_use]
pub fn gene_repertoire_diversity(pa: &[f64], n_genes: usize, n_genomes: usize) -> f64 {
    let mut sizes = vec![0_usize; n_genomes];
    for j in 0..n_genomes {
        let mut sum = 0;
        for i in 0..n_genes {
            if pa[i * n_genomes + j] > 0.5 {
                sum += 1;
            }
        }
        sizes[j] = sum;
    }

    let mut counts = std::collections::HashMap::new();
    for &s in &sizes {
        *counts.entry(s).or_insert(0_usize) += 1;
    }

    let total = n_genomes as f64;
    let freqs: Vec<f64> = counts.values().map(|&c| c as f64 / total).collect();
    primitives::shannon_entropy(&freqs)
}

/// Pairwise Jaccard distance between genomes (columns of PA matrix).
#[must_use]
pub fn jaccard_distance_matrix(pa: &[f64], n_genes: usize, n_genomes: usize) -> Vec<f64> {
    let mut dist = vec![0.0_f64; n_genomes * n_genomes];
    for i in 0..n_genomes {
        for j in (i + 1)..n_genomes {
            let mut intersection = 0.0_f64;
            let mut union = 0.0_f64;
            for g in 0..n_genes {
                let a = pa[g * n_genomes + i];
                let b = pa[g * n_genomes + j];
                intersection += a * b;
                union += a.max(b);
            }
            let d = if union > 0.0 {
                1.0 - intersection / union
            } else {
                0.0
            };
            dist[i * n_genomes + j] = d;
            dist[j * n_genomes + i] = d;
        }
    }
    dist
}

#[cfg(test)]
#[allow(clippy::float_cmp)]
mod tests {
    use super::*;
    use crate::tolerances;

    fn test_rng() -> Rng {
        Rng::new(42)
    }

    #[test]
    fn pa_matrix_is_binary() {
        let mut rng = test_rng();
        let env = vec![0, 0, 0, 1, 1, 1];
        let pa = generate_pa_matrix(6, 20, 0.25, 0.1, &mut rng, &env);
        assert!(pa.iter().all(|&v| v == 0.0 || v == 1.0));
    }

    #[test]
    fn partition_sums_to_total() {
        let freqs = vec![1.0, 1.0, 0.5, 0.5, 0.1];
        let (c, a, s) = partition_pangenome(&freqs, 10, 0.95);
        assert_eq!(c + a + s, 5);
    }

    #[test]
    fn neutral_spectrum_normalized() {
        let spec = neutral_spectrum(10);
        let sum: f64 = spec.iter().sum();
        assert!(
            (sum - 1.0).abs() < tolerances::CROSS_LANGUAGE,
            "neutral spectrum should sum to 1"
        );
    }

    #[test]
    fn chi_squared_zero_for_matching() {
        let obs = vec![10.0, 10.0, 10.0, 10.0];
        let exp = vec![0.25, 0.25, 0.25, 0.25];
        let chi2 = spectrum_chi_squared(&obs, &exp);
        assert!(chi2 < tolerances::CROSS_LANGUAGE);
    }

    #[test]
    fn jaccard_symmetric_and_bounded() {
        let pa = vec![1.0, 0.0, 1.0, 1.0, 0.0, 1.0];
        let dist = jaccard_distance_matrix(&pa, 2, 3);
        for i in 0..3 {
            assert!((dist[i * 3 + i]).abs() < tolerances::CROSS_LANGUAGE);
            for j in 0..3 {
                assert!((dist[i * 3 + j] - dist[j * 3 + i]).abs() < tolerances::CROSS_LANGUAGE);
                assert!(dist[i * 3 + j] >= 0.0 && dist[i * 3 + j] <= 1.0);
            }
        }
    }

    #[test]
    fn determinism() {
        let mut r1 = Rng::new(42);
        let mut r2 = Rng::new(42);
        let env = vec![0, 0, 1, 1];
        let pa1 = generate_pa_matrix(4, 10, 0.3, 0.1, &mut r1, &env);
        let pa2 = generate_pa_matrix(4, 10, 0.3, 0.1, &mut r2, &env);
        assert_eq!(pa1, pa2);
    }

    #[test]
    fn gene_frequencies_bounded() {
        let mut rng = Rng::new(42);
        let env = vec![0, 0, 1, 1];
        let pa = generate_pa_matrix(4, 20, 0.3, 0.1, &mut rng, &env);
        let freqs = gene_frequencies(&pa, 20, 4);
        assert_eq!(freqs.len(), 20);
        assert!(freqs.iter().all(|&f| (0.0..=1.0).contains(&f)));
    }

    #[test]
    fn frequency_spectrum_sums_correctly() {
        let freqs = vec![0.1, 0.3, 0.5, 0.7, 0.9, 0.2, 0.4, 0.6, 0.8, 0.95];
        let spec = frequency_spectrum(&freqs, 5);
        assert_eq!(spec.len(), 5);
        let total: f64 = spec.iter().sum();
        let expected = freqs.len() as f64;
        assert!(
            (total - expected).abs() < tolerances::ZERO_DETECTION,
            "frequency spectrum total {total} != expected {expected}"
        );
    }

    #[test]
    fn env_association_chi2_nonneg() {
        let mut rng = Rng::new(42);
        let env = vec![0, 0, 1, 1];
        let pa = generate_pa_matrix(4, 10, 0.3, 0.1, &mut rng, &env);
        let chi2 = env_association_chi2(&pa, 10, 4, &env);
        assert_eq!(chi2.len(), 10);
        assert!(chi2.iter().all(|&v| v >= 0.0));
    }

    #[test]
    fn selection_coefficient_nonneg() {
        let obs = vec![5.0, 10.0, 15.0, 12.0, 8.0];
        let neu = vec![0.2, 0.2, 0.2, 0.2, 0.2];
        let s = selection_coefficient(&obs, &neu);
        assert!(s >= 0.0 && s.is_finite());
    }

    #[test]
    fn gene_repertoire_diversity_positive() {
        let mut rng = Rng::new(42);
        let env = vec![0, 0, 1, 1];
        let pa = generate_pa_matrix(4, 20, 0.3, 0.1, &mut rng, &env);
        let d = gene_repertoire_diversity(&pa, 20, 4);
        assert!(d >= 0.0 && d.is_finite());
    }
}
